### PR TITLE
Removal of getPagesNumber() and other improvements

### PR DESCRIPTION
**Pull request by tempura-san**
@raulhaag This pull request is **not yet ready to be merged (as it is also untested)**, but is to be considered as basis for discussion and review until it is ripe for doing so.

I went through all the servers and removed the `getPagesNumber` function. Also single uses of `setExtra` were merged into the calling function (usually there remained only one callee).

I have made the following assumptions (please correct me if I am wrong):
* `chapterInit` is always called at least once before `getImageFrom` is called for the `Chapter` object.
* therefore there is no need to call it in `getImageFrom`.
* a `Chapter` object is considered initialised, as soon as the page count is more then zero (for avoiding double initialisation via `chapterInit`).

As there is still some redundant code I would like to remove, so I am asking for your opinion:
* the check for sane page numbers in `getImageFrom` could be avoided, if it is guaranteed, that the callee will **never** request page numbers outside of (1..page count).
* the check in `chapterInit` for initialisation could be avoided, if it is ensured, that the initialisation function for the chapter is only called **once** for each object.

Also various incorrect usages of `getFirstMatch` and `getFirstMatchDefault` were fixed. As the exception messages are displayed in the UI I have also changed them to translatable strings.

The french translation file already contained two strings from #460 which produce two warnings during build, as the code using them is not yet present - so they have been removed for now.

I have noticed, that there is a `get` function and a `getWithTimeout`. Which should be preferred? As of now, there is no feedback to the user, that the server connection timed out, so it might be misleading to assume that MiMangaNu is failing (e.g. for the filtered Manga list). Maybe this could be improved in the future.

It would be great to have proper unit tests for all the server implementations in order to quickly check for regressions and known bugs. Some basis seems to be present, but I have never done unit testing using Android Studio, so I guess I'll have to learn. :smile:

### CODE REVIEW

**Comment by raulhaag** [comment]
chapter init is called before getImageFrom thats is ok but you need to ensure that if the server use the extra field both field are setted (setExtra & setPages) or throw exception before save the changes of only one. 

getWhitTimeOut (the timeout can be set on settings cannection/read) that is usefull for bad and slow connections.

about test "AndroidTest" the basic made by me now need to set context for string reference

**Comment by tempura-san** [comment]
I have now removed the page number checking, as it is easy to forget. The calling function should be responsible for ensuring sanity.

I have also reordered the code to ensure `setPages` is called last. So only if all preceding instructions **did not yield an exception** the page number is finally set, indicating _chapter initialisation complete_.

Concerning `getWithTimeout`: How do you know if the connection is bad beforehand? Shouldn't this function be used by default?

**Comment by tempura-san** [comment]
Here is a first reworked version of the test suite for the servers - currently 8 servers out of 31 are failing the tests. Also I am not too satisfied with the test structure, so expect an update soon.

**Comment by raulhaag** [comment]
@tempura-san  about ninemanga i have the fix (not ready but i know wath to do) so when a other was ready i go to make it
getWithTimeout: you don't know user can set it in settings if needed (the normal get call to getWithTimeout and add all those options)
```java
public String get(String web) throws Exception {
        return this.get(web, connectionTimeout, writeTimeout, readTimeout);
    }
```

**Comment by tempura-san** [comment]
Ok - so it is redundant in this case. I have removed the `getWithTimeout()` function from the code and replaced its uses by `get()'.

Also I have fixed the workaround needed in the `TestServers` class for initialisation of the `Navigator.navigator` 'Singleton'. It is now replaced by a almost real Singleton.

In the test suite, now also the search results are tested (which increased the number of broken servers a bit).

**RawSenManga** seems to have had a complete redesign. The scraper had to be changed at some places to work again.
**MangaPedia** seems to deliver broken links ATM - can't be helped, does not work in the browser either.

For some other servers, only minor changed needed to be done.

Currently ~8 servers (mostly **NineManga**) out of the 31 fail the tests - so still some work to do...

**Comment by tempura-san** [comment]
Just an update. I have started to switch the Instrumentation API based `TestServers` to host based testing. This would enable faster testing times, without the need for a target device or emulator.
Also the server tests are not related to the device hardware, so it would be more fitting to do host-based test.

Currently I have parametrised tests using the Roboelectric framework running. Unfortunately I face two issues, which I have yet to sort out:
* the `KeyStore` on the host does not know of the `AndroidCAStore`. I have a proof-of-concept to fake this via the host's certificate store, but class shadowing does not yet work using Roboelectric.
* servers using the `Duktape` library report the library as missing - maybe some configuration issue.

Once these two issues are fixed, the host-based tests are on par with the current Instrumentation API tests. So please stay tuned.

**Comment by raulhaag** [comment]
Hi priority now is have the servers working, there is not a such of thing like a definitive update i really like your implementation and optimization, but need to make cut off  in time to time for releases not for me, for users that have a lot of servers down at this time (2 weeks or more), so after writing this, i go to do a release this Sunday.

**Comment by tempura-san** [comment]
I understand your point and therefore I try to work as fast as I can to get a good test framework to deliver a good experience to the end users.
Sunday sounds good. I'll try to fix the servers as good as possible for you to take over. Maybe the testing framework will then also be in better shape by then.

**Comment by tempura-san** [comment]
Here you go - all servers test now fine (except for **KuManga** which delivers broken links - which do not even show up in my browser...).
**NineManga** images are now loaded, but due to the hotlink detection it will just deliver the [bogus image](http://es.taadd.com/files/img/57ead05682694a7c026f99ad14abb8c1.jpg).
I could not test **Batoto**, as I do not have an account.

BTW: The tests are still using the Instrumentation API - for the time being.

Please check the changes and pull them in if they work properly. I'll open a new PR once I have further inputs.

PS: The CI build failed because it could not fetch some part (`rapid.decoder:jpeg-decoder:0.3.0` dependency) from GitHub - is there a way to start the build fresh, without having to push changes?

**Comment by raulhaag** [comment]
Ok don't worry about that i red the problem, merging

**Comment by raulhaag** [comment]
i just fix all ninemanga servers and all servers are working

**Comment by tempura-san** [comment]
Perfect! :+1:

**Comment by tempura-san** [comment]
Did you run the test suite on all of the servers? Just out of interest if everything is working as expected on your side as well.

**Comment by raulhaag** [comment]
yes tested, but still publish the 1.72 with errors (forget an !) just fixed a couple in v 1.73

**Comment by tempura-san** [comment]
I was about to add a test for this - but I am still struggling with the unit test framework. I have asked a question on the relevant issue of the Robolectric project [here](https://github.com/robolectric/robolectric/issues/1518).

**TL;DR:** The `java.lang.KeyStore` is not (yet properly) shadowed and fails on the host when trying to access the `AndroidCAStore`.

If all else fails, I could rewrite the `getTrustManagers` method in `Navigator` to check if it run on the device (use `AndroidCAStore`) or the unit test host (use host CA store).
